ufl/algorithms/compute_form_data.py

```diff
@@ -355,9 +358,19 @@ def compute_form_data(form,
     self.reduced_coefficients = sorted(reduced_coefficients_set,
                                        key=lambda c: c.count())
     self.num_coefficients = len(self.reduced_coefficients)
-    self.original_coefficient_positions = [i for i, c in enumerate(self.original_form.coefficients())
-                                           if c in self.reduced_coefficients]
 
+    updated_original_form_coefficients = self.original_form.coefficients()
+    self.coefficients = updated_original_form_coefficients
+    for c in self.original_form.coefficients():
+        if isinstance(c, ExternalOperator):
+            key_e = find_initial_external_operator(c)
+            updated_original_form_coefficients += tuple(e for e in ExternalOperator._ufl_all_external_operators_[key_e].values()
+                                                        if e in self.reduced_coefficients)
```

Comment:
Why do we have to inspect all possible external operators, and not just the ones in this form? This global dictionary is rather displeasing.

Reply:
Things were done in the assumption that the list of the coefficients in a given form could not change after having prepocessed the form, which is not the case here. Since when applying the differentiation rules we potentially create new coefficients (i.e the external operator’s derivatives). When we create the derivative of an external operator object we need to check if it has already been done or not to assign the correct counting, basically to say that in a given form the two derivatives E^{i1} and E^{i2} coming from a given external operator E are the same (refer to the same object) if they have the same derivatives multiindex (i1=i2), hence the presence of this weakreference dictionary to keep a trace of the existing external operators.

This weak referencing is not fully done yet, I will have to change some lines in tsfc_interface where the form compiling is done, since again there is behind the scenes an assumption that the coefficients do not change when compiling the form.